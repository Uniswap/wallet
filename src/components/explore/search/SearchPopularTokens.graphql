query SearchPopularTokens {
  topTokens(chain: ETHEREUM, orderBy: VOLUME, page: 1, pageSize: 3) {
    id
    address
    chain
    name
    symbol
    project {
      id
      logoUrl
    }
  }
  # `topTokens` returns WETH rather than ETH
  # here we retrieve ETH information to swap out in the UI
  eth: tokens(contracts: [{ address: null, chain: ETHEREUM }]) {
    id
    address
    chain
    name
    symbol
    project {
      id
      logoUrl
    }
  }
}
