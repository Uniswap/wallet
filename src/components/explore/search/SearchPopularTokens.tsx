import React, { useMemo } from 'react'
import { FlatList, ListRenderItemInfo } from 'react-native'
import { SearchTokenItem } from 'src/components/explore/search/items/SearchTokenItem'
import { Inset } from 'src/components/layout'
import { Loader } from 'src/components/loading'
import { ChainId } from 'src/constants/chains'
import { EMPTY_ARRAY } from 'src/constants/misc'
import { WRAPPED_NATIVE_CURRENCY } from 'src/constants/tokens'
import {
  Chain,
  SearchPopularTokensQuery,
  useSearchPopularTokensQuery,
} from 'src/data/__generated__/types-and-hooks'
import { SearchResultType, TokenSearchResult } from 'src/features/explore/searchHistorySlice'
import { areAddressesEqual } from 'src/utils/addresses'
import { fromGraphQLChain } from 'src/utils/chainId'
import { buildCurrencyId, buildNativeCurrencyId } from 'src/utils/currencyId'

export function SearchPopularTokens(): JSX.Element {
  // Load popular tokens by top Uniswap trading volume
  const { data, loading } = useSearchPopularTokensQuery()

  const popularTokens = useMemo(() => {
    if (!data || !data.topTokens) return EMPTY_ARRAY

    // special case to replace weth with eth because the backend does not return eth data
    // eth will be defined only if all the required data is available
    // when eth data is not fully available, we do not replace weth with eth
    const eth = data?.eth && data?.eth.length > 0 && data?.eth?.[0]?.project ? data.eth[0] : null
    const weth = WRAPPED_NATIVE_CURRENCY[ChainId.Mainnet]

    return data.topTokens
      .map((token) => {
        if (!token) return

        const isWeth =
          areAddressesEqual(token.address, weth.address) && token?.chain === Chain.Ethereum

        // manually replace weth with eth given backend only returns eth data as a proxy for eth
        if (isWeth && eth) {
          return gqlTokenToTokenSearchResult(eth)
        }

        return gqlTokenToTokenSearchResult(token)
      })
      .filter((t): t is TokenSearchResult => Boolean(t))
  }, [data])

  if (loading) {
    return (
      <Inset all="spacing8">
        <Loader.Token repeat={3} />
      </Inset>
    )
  }

  return (
    <FlatList
      data={popularTokens}
      keyExtractor={tokenKey}
      listKey="tokens"
      renderItem={renderTokenItem}
    />
  )
}

function gqlTokenToTokenSearchResult(
  token: NullUndefined<NonNullable<NonNullable<SearchPopularTokensQuery['topTokens']>[0]>>
): TokenSearchResult | null {
  if (!token || !token.project) return null

  const { chain, address, symbol, name, project } = token
  const chainId = fromGraphQLChain(chain)
  if (!chainId || !symbol || !name) return null

  return {
    type: SearchResultType.Token,
    chainId,
    address,
    name,
    symbol,
    logoUrl: project?.logoUrl,
  } as TokenSearchResult
}

const renderTokenItem = ({ item }: ListRenderItemInfo<TokenSearchResult>): JSX.Element => (
  <SearchTokenItem token={item} />
)

const tokenKey = (token: TokenSearchResult): string => {
  return token.address
    ? buildCurrencyId(token.chainId, token.address)
    : buildNativeCurrencyId(token.chainId)
}
