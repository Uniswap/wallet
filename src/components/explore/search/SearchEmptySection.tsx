import React from 'react'
import { useTranslation } from 'react-i18next'
import { FlatList, ListRenderItemInfo } from 'react-native'
import { FadeIn, FadeOut } from 'react-native-reanimated'
import { useAppDispatch, useAppSelector } from 'src/app/hooks'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { SearchEtherscanItem } from 'src/components/explore/search/items/SearchEtherscanItem'
import { SearchTokenItem } from 'src/components/explore/search/items/SearchTokenItem'
import { SearchWalletItem } from 'src/components/explore/search/items/SearchWalletItem'
import { SearchPopularTokens } from 'src/components/explore/search/SearchPopularTokens'
import { AnimatedFlex, Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import {
  clearSearchHistory,
  SearchResult,
  SearchResultType,
  selectSearchHistory,
  WalletCategory,
  WalletSearchResult,
} from 'src/features/explore/searchHistorySlice'

export const SUGGESTED_WALLETS: WalletSearchResult[] = [
  {
    type: SearchResultType.Wallet,
    address: '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',
    ensName: 'vitalik.eth',
    category: WalletCategory.Builder,
  },
  {
    type: SearchResultType.Wallet,
    address: '0x50EC05ADe8280758E2077fcBC08D878D4aef79C3',
    ensName: 'hayden.eth',
    category: WalletCategory.Builder,
  },
]

export function SearchEmptySection(): JSX.Element {
  const { t } = useTranslation()
  const dispatch = useAppDispatch()
  const searchHistory = useAppSelector(selectSearchHistory)

  const onPressClearSearchHistory = (): void => {
    dispatch(clearSearchHistory())
  }

  // Show search history (if applicable), trending tokens, and wallets
  return (
    <AnimatedFlex entering={FadeIn} exiting={FadeOut} gap="spacing12">
      {searchHistory.length > 0 && (
        <AnimatedFlex entering={FadeIn} exiting={FadeOut}>
          <FlatList
            ListHeaderComponent={
              <Flex
                row
                alignItems="center"
                justifyContent="space-between"
                mb="spacing4"
                ml="spacing8">
                <Text color="textSecondary" variant="subheadSmall">
                  {t('Recent searches')}
                </Text>
                <TouchableArea onPress={onPressClearSearchHistory}>
                  <Text color="accentAction" variant="buttonLabelSmall">
                    {t('Clear all')}
                  </Text>
                </TouchableArea>
              </Flex>
            }
            data={searchHistory}
            renderItem={renderSearchHistoryItem}
          />
        </AnimatedFlex>
      )}
      <Flex gap="spacing4">
        <Text color="textSecondary" mx="spacing8" variant="subheadSmall">
          {t('Popular tokens')}
        </Text>
        <SearchPopularTokens />
      </Flex>
      <FlatList
        ListHeaderComponent={
          <Text color="textSecondary" mb="spacing4" mx="spacing8" variant="subheadSmall">
            {t('Suggested wallets')}
          </Text>
        }
        data={SUGGESTED_WALLETS}
        keyExtractor={walletKey}
        listKey="wallets"
        renderItem={renderWalletItem}
      />
    </AnimatedFlex>
  )
}

const renderSearchHistoryItem = ({
  item: searchResult,
}: ListRenderItemInfo<SearchResult>): JSX.Element => {
  if (searchResult.type === SearchResultType.Token) {
    return <SearchTokenItem token={searchResult} />
  } else if (searchResult.type === SearchResultType.Wallet) {
    return <SearchWalletItem wallet={searchResult} />
  } else {
    return <SearchEtherscanItem etherscanResult={searchResult} />
  }
}

const renderWalletItem = ({ item }: ListRenderItemInfo<WalletSearchResult>): JSX.Element => (
  <SearchWalletItem wallet={item} />
)

const walletKey = (wallet: WalletSearchResult): string => {
  return wallet.address
}
