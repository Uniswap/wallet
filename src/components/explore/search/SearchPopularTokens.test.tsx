import React from 'react'
import { SearchPopularTokens } from 'src/components/explore/search/SearchPopularTokens'
import { SearchPopularTokensDocument } from 'src/data/__generated__/types-and-hooks'
import { EthToken, TopTokens } from 'src/test/gqlFixtures'
import { render, screen } from 'src/test/test-utils'

const mock = {
  request: {
    query: SearchPopularTokensDocument,
  },
  result: {
    data: {
      topTokens: TopTokens,
      eth: EthToken,
    },
  },
}

describe(SearchPopularTokens, () => {
  it('renders without error', async () => {
    const tree = render(<SearchPopularTokens />, { mocks: [mock] })

    // Loading should show Token loader
    expect(screen.getAllByText('Token Full Name')).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()

    // Success where WETH result in topTokens is replaced by ETH
    expect(await screen.findByText('ETH')).toBeDefined()
    expect(screen.getByText('WBTC')).toBeDefined()
    expect(screen.getByText('USDC')).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
