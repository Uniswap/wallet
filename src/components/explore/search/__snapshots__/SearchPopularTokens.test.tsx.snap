// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPopularTokens renders without error 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "padding": 8,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "display": "flex",
            "flexDirection": "column",
            "flexGrow": 1,
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "minHeight": 56,
              "opacity": 1,
              "paddingVertical": 8,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "flexShrink": 1,
                "justifyContent": "flex-start",
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#E8ECFB",
                  "borderRadius": 999999,
                  "display": "flex",
                  "minHeight": 40,
                  "minWidth": 40,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 12,
                  "width": 12,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "flexShrink": 1,
                  "justifyContent": "flex-start",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                      "position": "relative",
                    },
                  ]
                }
              >
                <View
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "color": "transparent",
                          "display": "flex",
                          "fontFamily": "Inter-Medium",
                          "fontSize": 17,
                          "lineHeight": 24,
                          "opacity": 0,
                        },
                      ]
                    }
                  >
                    Token Full Name
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#E8ECFB",
                        "borderRadius": 4,
                        "bottom": "5%",
                        "display": "flex",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": "5%",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 0,
                  },
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "minHeight": 20,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "flexDirection": "row",
                        "position": "relative",
                      },
                    ]
                  }
                >
                  <View
                    accessibilityElementsHidden={true}
                    importantForAccessibility="no-hide-descendants"
                  >
                    <Text
                      allowFontScaling={true}
                      maxFontSizeMultiplier={1.4}
                      numberOfLines={1}
                      style={
                        Array [
                          Object {
                            "color": "transparent",
                            "display": "flex",
                            "fontFamily": "Inter-Medium",
                            "fontSize": 15,
                            "lineHeight": 20,
                            "opacity": 0,
                          },
                        ]
                      }
                    >
                      1,000 TFN
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "backgroundColor": "#E8ECFB",
                          "borderRadius": 4,
                          "bottom": "5%",
                          "display": "flex",
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": "5%",
                        },
                      ]
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexGrow": 0,
              "flexShrink": 0,
              "height": 4,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "minHeight": 56,
              "opacity": 0.6666666666666666,
              "paddingVertical": 8,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "flexShrink": 1,
                "justifyContent": "flex-start",
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#E8ECFB",
                  "borderRadius": 999999,
                  "display": "flex",
                  "minHeight": 40,
                  "minWidth": 40,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 12,
                  "width": 12,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "flexShrink": 1,
                  "justifyContent": "flex-start",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                      "position": "relative",
                    },
                  ]
                }
              >
                <View
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "color": "transparent",
                          "display": "flex",
                          "fontFamily": "Inter-Medium",
                          "fontSize": 17,
                          "lineHeight": 24,
                          "opacity": 0,
                        },
                      ]
                    }
                  >
                    Token Full Name
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#E8ECFB",
                        "borderRadius": 4,
                        "bottom": "5%",
                        "display": "flex",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": "5%",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 0,
                  },
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "minHeight": 20,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "flexDirection": "row",
                        "position": "relative",
                      },
                    ]
                  }
                >
                  <View
                    accessibilityElementsHidden={true}
                    importantForAccessibility="no-hide-descendants"
                  >
                    <Text
                      allowFontScaling={true}
                      maxFontSizeMultiplier={1.4}
                      numberOfLines={1}
                      style={
                        Array [
                          Object {
                            "color": "transparent",
                            "display": "flex",
                            "fontFamily": "Inter-Medium",
                            "fontSize": 15,
                            "lineHeight": 20,
                            "opacity": 0,
                          },
                        ]
                      }
                    >
                      1,000 TFN
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "backgroundColor": "#E8ECFB",
                          "borderRadius": 4,
                          "bottom": "5%",
                          "display": "flex",
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": "5%",
                        },
                      ]
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexGrow": 0,
              "flexShrink": 0,
              "height": 4,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "minHeight": 56,
              "opacity": 0.3333333333333333,
              "paddingVertical": 8,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "flexShrink": 1,
                "justifyContent": "flex-start",
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#E8ECFB",
                  "borderRadius": 999999,
                  "display": "flex",
                  "minHeight": 40,
                  "minWidth": 40,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 12,
                  "width": 12,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "flexShrink": 1,
                  "justifyContent": "flex-start",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                      "position": "relative",
                    },
                  ]
                }
              >
                <View
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "color": "transparent",
                          "display": "flex",
                          "fontFamily": "Inter-Medium",
                          "fontSize": 17,
                          "lineHeight": 24,
                          "opacity": 0,
                        },
                      ]
                    }
                  >
                    Token Full Name
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#E8ECFB",
                        "borderRadius": 4,
                        "bottom": "5%",
                        "display": "flex",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": "5%",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 0,
                  },
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "minHeight": 20,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "flexDirection": "row",
                        "position": "relative",
                      },
                    ]
                  }
                >
                  <View
                    accessibilityElementsHidden={true}
                    importantForAccessibility="no-hide-descendants"
                  >
                    <Text
                      allowFontScaling={true}
                      maxFontSizeMultiplier={1.4}
                      numberOfLines={1}
                      style={
                        Array [
                          Object {
                            "color": "transparent",
                            "display": "flex",
                            "fontFamily": "Inter-Medium",
                            "fontSize": 15,
                            "lineHeight": 20,
                            "opacity": 0,
                          },
                        ]
                      }
                    >
                      1,000 TFN
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "backgroundColor": "#E8ECFB",
                          "borderRadius": 4,
                          "bottom": "5%",
                          "display": "flex",
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": "5%",
                        },
                      ]
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`SearchPopularTokens renders without error 2`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "address": null,
        "chainId": 1,
        "logoUrl": "ethlogo.png",
        "name": "Ethereum",
        "symbol": "ETH",
        "type": 1,
      },
      Object {
        "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
        "chainId": 1,
        "logoUrl": "wbtclogo.png",
        "name": "Wrapped Bitcoin",
        "symbol": "WBTC",
        "type": 1,
      },
      Object {
        "address": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
        "chainId": 1,
        "logoUrl": "usdclogo.png",
        "name": "USD Coin",
        "symbol": "USDC",
        "type": 1,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  listKey="tokens"
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 5,
            "left": 5,
            "right": 5,
            "top": 5,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "display": "flex",
            "opacity": 1,
          }
        }
        testID="search-token-item"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "paddingHorizontal": 8,
                "paddingVertical": 12,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 40,
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "uri": "ethlogo.png",
                }
              }
              style={
                Array [
                  Object {
                    "resizeMode": "contain",
                  },
                  Object {
                    "backgroundColor": "#D2D9EE",
                    "borderColor": "#D2D9EE",
                    "borderRadius": 20,
                    "borderWidth": 0.5,
                    "height": 40,
                    "width": 40,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 12,
                  "width": 12,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "stretch",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  },
                ]
              }
            >
              <Text
                allowFontScaling={true}
                maxFontSizeMultiplier={1.4}
                style={
                  Array [
                    Object {
                      "color": "#0E1524",
                      "display": "flex",
                      "fontFamily": "Inter-Medium",
                      "fontSize": 17,
                      "lineHeight": 24,
                    },
                  ]
                }
              >
                Ethereum
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 0,
                  },
                ]
              }
            />
            <Text
              allowFontScaling={true}
              maxFontSizeMultiplier={1.4}
              style={
                Array [
                  Object {
                    "color": "#5D6785",
                    "display": "flex",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 15,
                    "lineHeight": 20,
                  },
                ]
              }
            >
              ETH
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 5,
            "left": 5,
            "right": 5,
            "top": 5,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "display": "flex",
            "opacity": 1,
          }
        }
        testID="search-token-item"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "paddingHorizontal": 8,
                "paddingVertical": 12,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 40,
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "uri": "wbtclogo.png",
                }
              }
              style={
                Array [
                  Object {
                    "resizeMode": "contain",
                  },
                  Object {
                    "backgroundColor": "#D2D9EE",
                    "borderColor": "#D2D9EE",
                    "borderRadius": 20,
                    "borderWidth": 0.5,
                    "height": 40,
                    "width": 40,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 12,
                  "width": 12,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "stretch",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  },
                ]
              }
            >
              <Text
                allowFontScaling={true}
                maxFontSizeMultiplier={1.4}
                style={
                  Array [
                    Object {
                      "color": "#0E1524",
                      "display": "flex",
                      "fontFamily": "Inter-Medium",
                      "fontSize": 17,
                      "lineHeight": 24,
                    },
                  ]
                }
              >
                Wrapped Bitcoin
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 0,
                  },
                ]
              }
            />
            <Text
              allowFontScaling={true}
              maxFontSizeMultiplier={1.4}
              style={
                Array [
                  Object {
                    "color": "#5D6785",
                    "display": "flex",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 15,
                    "lineHeight": 20,
                  },
                ]
              }
            >
              WBTC
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 5,
            "left": 5,
            "right": 5,
            "top": 5,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "display": "flex",
            "opacity": 1,
          }
        }
        testID="search-token-item"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "paddingHorizontal": 8,
                "paddingVertical": 12,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 40,
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "uri": "usdclogo.png",
                }
              }
              style={
                Array [
                  Object {
                    "resizeMode": "contain",
                  },
                  Object {
                    "backgroundColor": "#D2D9EE",
                    "borderColor": "#D2D9EE",
                    "borderRadius": 20,
                    "borderWidth": 0.5,
                    "height": 40,
                    "width": 40,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 12,
                  "width": 12,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "stretch",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  },
                ]
              }
            >
              <Text
                allowFontScaling={true}
                maxFontSizeMultiplier={1.4}
                style={
                  Array [
                    Object {
                      "color": "#0E1524",
                      "display": "flex",
                      "fontFamily": "Inter-Medium",
                      "fontSize": 17,
                      "lineHeight": 24,
                    },
                  ]
                }
              >
                USD Coin
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 0,
                  },
                ]
              }
            />
            <Text
              allowFontScaling={true}
              maxFontSizeMultiplier={1.4}
              style={
                Array [
                  Object {
                    "color": "#5D6785",
                    "display": "flex",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 15,
                    "lineHeight": 20,
                  },
                ]
              }
            >
              USDC
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
